src/sentry/spans/buffer/redis.py

```diff
@@ -30,6 +47,74 @@ class RedisSpansBuffer:
     def __init__(self):
         self.client: RedisCluster | StrictRedis = get_redis_client()
 
+    def batch_write_and_check_processing(
+        self,
+        spans_map: Mapping[SegmentKey, list[bytes]],
+        segment_first_seen_ts: Mapping[SegmentKey, int],
+        latest_ts_by_partition: Mapping[int, int],
+    ) -> list[ProcessSegmentsContext]:
+        """
+        1. Pushes batches of spans to redis
+        2. Check if number of spans pushed == to the number of elements that exist on the key. This
+            tells us if it was the first time we see the key. This works fine because RPUSH is atomic.
+        3. If it is the first time we see a particular segment, push the segment id and first seen
+            timestamp to a bucket so we know when it is ready to be processed.
+        3. Checks if 1 second has passed since the last time segments were processed for a partition.
+        """
+        keys = list(spans_map.keys())
+        spans_written_per_segment = []
+
+        # Batch write spans in a segment
+        with self.client.pipeline() as p:
+            for key in keys:
+                segment_id, project_id, partition = key
+                spans = spans_map[key]
+                segment_key = get_segment_key(project_id, segment_id)
+                # RPUSH is atomic
+                p.rpush(segment_key, *spans)
+                spans_written_per_segment.append(len(spans))
+
+            results = p.execute()
+
+        partitions = list(latest_ts_by_partition.keys())
+        with self.client.pipeline() as p:
+            # Get last processed timestamp for each partition processed
+            # by consumer
+            for partition in partitions:
+                timestamp = latest_ts_by_partition[partition]
+                timestamp_key = get_last_processed_timestamp_key(partition)
+                # GETSET is atomic
+                p.getset(timestamp_key, timestamp)
+
+            for result in zip(keys, spans_written_per_segment, results):
+                # Check if this is a new segment, if yes, add to bucket to be processed
+                key, num_written, num_total = result
+                if num_written == num_total:
+                    segment_id, project_id, partition = key
+                    segment_key = get_segment_key(project_id, segment_id)
+                    bucket = get_unprocessed_segments_key(partition)
+
+                    timestamp = segment_first_seen_ts[key]
+                    p.expire(segment_key, SEGMENT_TTL)
+                    p.rpush(bucket, json.dumps([timestamp, segment_key]))
```

Comment:
Do we have to push JSON here? Can't we push the values directly and perhaps have odd index be timestamps, even be segment keys? Saving on serialization/deserialization for every key would help.

Reply:
https://github.com/getsentry/sentry/pull/69445/commits/123238a21e1d537867068577f2a1fd2679ce3bf6
created a new key since we don't want to push to the existing bucket. but that would mean we wouldn't process the segments that are currently stored on the old key (so miss about 2 min worth of segments, which is fine, and they would just expire when they hit their TTL - but we'd have to expire all the old buckets still). saves us a json.dumps, but we do end up having to parse the results since we return force return bytes, is it worth it?

Reply:
I think that's worth it yes.